src/odemis/driver/autoscript_client.py

```diff
@@ -0,0 +1,2304 @@
+    def start_generate(self):
+        self._detector.start_generate()
+
+    def stop_generate(self):
+        self._detector.stop_generate()
+
+
+class Stage(model.Actuator):
+    """
+    This is an extension of the model.Actuator class. It provides functions for
+    moving the TFS stage and updating the position.
+    """
+
+    def __init__(self, name, role, parent, rng=None, **kwargs):
+        if rng is None:
+            rng = {}
+        stage_info = parent.stage_info()
+        if "x" not in rng:
+            rng["x"] = stage_info["range"]["x"]
+        if "y" not in rng:
+            rng["y"] = stage_info["range"]["y"]
+        if "z" not in rng:
+            rng["z"] = stage_info["range"]["z"]
+        if "rx" not in rng:
+            rng["rx"] = stage_info["range"]["T"]
+        if "rz" not in rng:
+            rng["rz"] = stage_info["range"]["R"]
+
+        axes_def = {
+            "x": model.Axis(unit=stage_info["unit"]["x"], range=rng["x"]),
+            "y": model.Axis(unit=stage_info["unit"]["y"], range=rng["y"]),
+            "z": model.Axis(unit=stage_info["unit"]["z"], range=rng["z"]),
+            "rx": model.Axis(unit=stage_info["unit"]["T"], range=rng["rx"]),
+            "rz": model.Axis(unit=stage_info["unit"]["R"], range=rng["rz"]),
+        } # TODO: make these axis and unit arguments consistent
+
+        model.Actuator.__init__(self, name, role, parent=parent, axes=axes_def,
+                                **kwargs)
+        # will take care of executing axis move asynchronously
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+
+        self.position = model.VigilantAttribute({}, unit=stage_info["unit"],
+                                                readonly=True)
+        self._updatePosition()
+
+        # Refresh regularly the position
+        self._pos_poll = util.RepeatingTimer(5, self._refreshPosition, "Stage position polling")
+        self._pos_poll.start()
+
+    def _updatePosition(self):
+        """
+        update the position VA
+        """
+        old_pos = self.position.value
+        pos = self._getPosition()
+        self.position._set_value(self._applyInversion(pos), force_write=True)
+        if old_pos != self.position.value:
+            logging.debug("Updated position to %s", self.position.value)
+
+    def _refreshPosition(self):
+        """
+        Called regularly to update the current position
+        """
+        # We don't use the VA setters, to avoid sending back to the hardware a
+        # set request
+        logging.debug("Updating SEM stage position")
+        try:
+            self._updatePosition()
+        except Exception:
+            logging.exception("Unexpected failure when updating position")
+
+    def _getPosition(self):
+        """Get position and translate the axes names to be Odemis compatible."""
+        pos = self.parent.get_stage_position()
+        pos["rx"] = pos.pop("t")
+        pos["rz"] = pos.pop("r")
```

Comment:
uppercase or lowercase t & r? In the init it's uppercase... It might be worthy to add a comment explaining it's important.

Reply:
api returns lowercase r, t for stage position, and uppercase for axis limits. I've changed it so the server will now return lowercase for all